### ISSUE TITLE
How to warm up a system followed by NVT/NPT equilibration while slowly removing backbone constraints

### ISSUE THREAD

**Issue by noahharrison64** [openmm/openmm#3782]
Hi,

I'm attempting to run some of my first simulations using OpenMM. I want to use the following production pipeline:
After applying backbone restraints
1) Energy minimize
2) Slowly warm the system up under constant volume (NVT equilibration)
3) Slowly remove the backbone restraints while equilbrating under NPT
4) Run a 10ns production run

I have struggled to figure out how to do steps 2 and 3. Could someone check whether the below code looks appropriate? 

```py
# Function to add backbone position restraints
def add_backbone_posres(system, positions, atoms, restraint_force):
  force = CustomExternalForce("k*periodicdistance(x, y, z, x0, y0, z0)^2")
  force_amount = restraint_force * kilocalories_per_mole/angstroms**2
  force.addGlobalParameter("k", force_amount)
  force.addPerParticleParameter("x0")
  force.addPerParticleParameter("y0")
  force.addPerParticleParameter("z0")
  for i, (atom_crd, atom) in enumerate(zip(positions, atoms)):
    if atom.name in  ('CA', 'C', 'N'):
      force.addParticle(i, atom_crd.value_in_unit(nanometers))
  posres_sys = deepcopy(system)
  posres_sys.addForce(force)
  return posres_sys

# Set up 
pdb = PDBFile(fixed_pdb_file)
forcefield = ForceField('amber99sb.xml', 'tip3p.xml')
modeller = Modeller(pdb.topology, pdb.positions)
modeller.addSolvent(forcefield, padding=1.0*nanometers)

#System preperation 
system = forcefield.createSystem(modeller.topology, nonbondedMethod=app.PME,
    nonbondedCutoff=1.0*nanometers, constraints=app.HBonds, rigidWater=True,
    ewaldErrorTolerance=0.0005)
posres_sys = add_backbone_posres(system, modeller.positions, modeller.topology.atoms(), 100)
integrator = LangevinIntegrator(300*kelvin, 1/picosecond, 0.002*picoseconds)
integrator.setConstraintTolerance(0.00001)
simulation = Simulation(modeller.topology, posres_sys, integrator)
simulation.reporters.append(
    StateDataReporter(
        stdout,
        10000,
        step=True,
        potentialEnergy=True,
        temperature=True,
        volume=True,
        density=True
    )
)

#Minimize
print('Minimizing...')
simulation.context.setPositions(modeller.positions)
simulation.minimizeEnergy()
simulation.step(3500)

#WarmUp with a NVT run.  Slowly warm up temperature - every 1000 steps raise the temperature by 5 K, ending at 300 K
simulation.context.setVelocitiesToTemperature(5*kelvin)
print('Warming up the system...')
T = 5
mdsteps = 50000
for i in range(60):
  simulation.step(int(mdsteps/60) )
  temperature = (T+(i*T))*kelvin 
  integrator.setTemperature(temperature)


#NPT equilibration, reducing backbone constraints
mdsteps = 500000
barostat = system.addForce(MonteCarloBarostat(1*atmosphere, 300*kelvin))
simulation.context.reinitialize(True)
print('Running NPT equilibration...')
for i in range(100):
  simulation.step(int(mdsteps/100))
  simulation.context.setParameter('k', (float(99.02-(i*0.98))*kilocalories_per_mole/angstroms**2))

# save the equilibration results to file : state is platform independent but less precise, checkpoint file
simulation.saveState((workDir/'eq.state').as_posix())
simulation.saveCheckpoint((workDir/'eq.chk').as_posix())

# Load checkpoint
# reset step and time counters
simulation.loadCheckpoint((workDir/'eq.chk').as_posix())
eq_state = simulation.context.getState(getVelocities=True, getPositions=True)
positions = eq_state.getPositions()
velocities = eq_state.getVelocities()
integrator = LangevinIntegrator(300*kelvin, 1/picosecond, 0.002*picoseconds)
integrator.setConstraintTolerance(0.00001)
simulation = Simulation(modeller.topology, system, integrator)
simulation.context.setPositions(positions)
simulation.context.setVelocities(velocities)

# append reporters
simulation.reporters.append(
    DCDReporter((workDir/'10ns_traj.dcd').as_posix(), 1000))
simulation.reporters.append(
    StateDataReporter(
         stdout,
         10000,
         step=True,
         potentialEnergy=True,
         temperature=True,
         progress=True,
         remainingTime=True,
        speed=True,
        totalSteps=mdsteps,
        separator='\t'
        )
    )
simulation.reporters.append(
    PDBReporter(
        (workDir/'10ns_production_run_snapshots.pdb').as_posix(),
        reportInterval = 10000
    )
)

# run production simulation fpr 10ns
mdsteps = 20000000
print('Running Production...')
simulation.step(mdsteps)
print('Done!')
simulation.saveState((workDir/'sim_10ns.state').as_posix())
simulation.saveCheckpoint((workDir/'sim_10ns.chk').as_posix())

```

**Comment by peastman** [comment]
That mostly looks fine, at least on a quick read through.  You don't reduce the restraint force quite all the way to zero.  I think you want to add

```python
simulation.context.setParameter('k', 0)
```

before starting the production part.

**Comment by peastman** [comment]
That's normal.  See https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#periodic for a discussion of it.

**Comment by noahharrison64** [comment]
Okay great thank you Peter. 
Last question (Hopefully) - If I'm using a Langevin integrator (and no barostat) does that mean I'm running an NVT simulation? My understanding is yes, because a Langevin integrator is defined as:
_LangevinIntegator simulates a system in contact with a heat bath by integrating the Langevin equation of motion:_

And then by adding a barostat I can carry out an NPT simulation? 
And if I were to have no barostat, and an integrator which isn't simulating a system coupled to a heatbath (as defined [here](http://docs.openmm.org/latest/userguide/theory/04_integrators.html) then I would be running an NVE simulation?

**Comment by peastman** [comment]
Exactly.

**Comment by noahharrison64** [comment]
Is it possible to remove a barostat from a system once having run an NPT simulation?

**Comment by blebleblesm** [comment]
I tried to implement this equilibration protocol but I'm struggling with it so hopefully it's okay if I chime in.
So, I realized that, as far as I understand, the barostat isn't applied to the right system during the NPT simulation:
`system.addForce(MonteCarloBarostat(1*atmosphere, 300*kelvin))`
should be
`posres_sys.addForce(MonteCarloBarostat(1*atmosphere, 300*kelvin))`
or am I mistaken here?

Applying this change, however the production simulation is blowing up and giving an "Particle coordinate is nan" exception.

I'm at a loss how to proceed from here, I tried to use setFrequency to turn off the barostat at the end of the equilibration but that also didn't help.
